### ISSUE TITLE
`--report-plist` broken in current master

### ISSUE THREAD

**Issue by eahrold** [autopkg/autopkg#163]
Not sure if this is on anyone's radar, but the switch to table style result (https://github.com/autopkg/autopkg/commit/d688ed6aac79ca5c68b801aac7b83acc6ce1da5e) broke the `--report-plist` feature.

```
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 1634, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 1628, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 1558, in run_recipes
    results_report[event_type] = locals()[event_type]
KeyError: 'new_downloads'
```

**Comment by gregneagle** [comment]
This functionality will have to be completely rethought and rewritten; I want to move away from the hard-coded special event types and allow processors to define these themselves.

I never have used that functionality -- but I suppose Autopkgr does...

**Comment by eahrold** [comment]
I agree that makes much more sense. And as far as I'm concerned the report.plist could have any key-values, even just using the `summary_results` may work just fine.

**Comment by timsutton** [comment]
Took a quick look at how `summary_results` would look as-is:

```plist
    <key>summaries</key>
    <dict>
        <key>pkg_creator_summary_result</key>
        <dict>
            <key>header</key>
            <string>Identifier               Version          Pkg path
----------               -------          --------                        </string>
            <key>rows</key>
            <array>
                <string>com.adiumX.adiumX.pkg    1.5.10           /Users/tsutton/Library/AutoPkg/Cache/com.github.autopkg.pkg.Adium/Adium-1.5.10.pkg</string>
            </array>
            <key>summary_text</key>
            <string>The following packages were built:</string>
        </dict>
        <key>url_downloader_summary_result</key>
        <dict>
            <key>header</key>
            <string></string>
            <key>rows</key>
            <array>
                <string>/Users/tsutton/Library/AutoPkg/Cache/com.github.autopkg.pkg.Adium/downloads/Adium.dmg</string>
            </array>
            <key>summary_text</key>
            <string>The following new items were downloaded:</string>
        </dict>
    </dict>
```

So yes, I think there's a need to let processors set a more structured format.

**Comment by gregneagle** [comment]
As I mentioned to Tim on IRC, the direction I'm heading is to tear out the report printing code from the autopkg tool and instead have the Processors that declare interesting items to report on also declare a function for printing them.

We can then have the plist contain the data without stuffing formatting in there as well.

**Comment by timsutton** [comment]
I also thought that it makes sense to have an abstract report-printing function (handling the headings and spacing, etc.), either in `autopkg` or some module within `autopkglib`, but wouldn't it be redundant to have multiple processors each implementing a printing function?

Wouldn't this be more the responsibility of the core tool to still handle the actual reporting/display, even though each processor can declare the interesting item data, the names for them, "description of what these items are" heading, etc.?

**Comment by gregneagle** [comment]
Each processor can decide what to store/report on. In the case of URLDownloader, we track a single string listing the item downloaded. But other processors have arbitrary things they track: filenames, versions, catalogs, identifiers, etc. It would be incredibly difficult to write a generic report printer that could  handle those. The current code cheats by embedding formatting in the tracked items, but that makes the data less useful overall.

The previous code cheated a different way by being hardwired to a few summary types, and could not/did not report on things like install recipes, jss recipes, etc, since none of that was in the (hard) code.

If instead, the Processor defines a print_summary_report type function, it can know its own special needs (and its own data) and print it appropriately.

**Comment by eahrold** [comment]
You could potentially bring in the header/rows as arrays representing columns from the processors which could then be constructed like this...

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>failures</key>
	<array/>
	<key>results</key>
	<dict>
		<key>pkg_creator_summary_result</key>
		<dict>
			<key>header</key>
			<array>
				<string>Identifier</string>
				<string>Version</string>
				<string>Pkg path</string>
			</array>
			<key>rows</key>
			<array>
				<array>
					<string>com.adiumX.adiumX.pkg</string>
					<string>1.5.10</string>
					<string>~/Library/AutoPkg/Cache/com.github.autopkg.pkg.Adium/Adium-1.5.10.pkg</string>
				</array>
			</array>
			<key>summary_text</key>
			<string>The following packages were built:</string>
		</dict>
		<key>url_downloader_summary_result</key>
		<dict>
			<key>header</key>
			<array/>
			<key>rows</key>
			<array>
				<array>
					<string>~/Library/AutoPkg/Cache/com.github.autopkg.pkg.Adium/downloads/Adium.dmg</string>
				</array>
			</array>
			<key>summary_text</key>
			<string>The following new items were downloaded:</string>
		</dict>
	</dict>
</dict>
</plist>
```

**Comment by timsutton** [comment]
Ok, I see that argument about the processor having more specific needs than AutoPkg may be able to generalize for, at least as far as report display goes. I was going under an assumption that one wouldn't typically have more than 3 or 4 things to summarize for a given run of a processor, and that it would be possible to have a generic enough printing function.

I imagine there might be a case where the processor might want to record some additional data in the report plist that doesn't make sense (or is too dense) for display, but that could be useful to systems collecting the data.

**Comment by eahrold** [comment]
@timsutton, @gregneagle, I wanted to check back in and see if this is something either of you were working on. I've put together a solution that basically converts the `_summary_results` to a plist as mentioned [above](https://github.com/autopkg/autopkg/issues/163#issuecomment-79085994), but I won't bother putting up a PR if you've got a more comprehensive approach in mind.

Here's what I have
```Python
    if options.report_plist:
        report_results = {}
        report = { 'failures': failures }

    if summary_results:
        for key, value in summary_results.items():
            log("\n%s" % value['summary_text'])
            if options.report_plist:
                report_results[key] = {}
                report_results[key]['summary_text'] = value['summary_text']
                report_results[key]['rows'] = []

            header = value.get('header')
            if header:
                log("    %s" % header)
                # Create visual underline of header columns
                log("    %s" % re.sub(r'(\w)', '-', header))

                if options.report_plist:
                    # Only split when whitespace is 2 or more.
                    report_results[key]['header'] = re.split(r'\s{2,}', header)
            for row in value['rows']:
                log("    %s" % row)
                if options.report_plist:
                    report_results[key]['rows'].append(re.split(r'\s{2,}', row))

    if not summary_results:
        log("\nNothing downloaded, packaged or imported.")

    # save report plist with the summary data
    if options.report_plist:
        report = {'failures': failures, 'results': report_results}
        write_plist_exit_on_fail(report, options.report_plist)
        log("\nReport plist saved to %s." % options.report_plist)
```

This also requires adjusting the the _summary_result's headers key in MunkiImporter.py and PkgCreator.py making them single line and handling the column decoration here, but that's the only other change.

Thanks.

**Comment by gregneagle** [comment]
This still feels wrong to me. Too much hard-coded presentation info in the data. We need to abstract this further.

**Comment by gregneagle** [comment]
I've pushed another approach: https://github.com/autopkg/autopkg/commit/7e210af12286952a8027d8adb528b4c79720020b

**Comment by gregneagle** [comment]
I'm not too excited about repeating the field names like here: https://github.com/autopkg/autopkg/blob/7e210af12286952a8027d8adb528b4c79720020b/Code/autopkglib/PkgCreator.py#L189-L197

But since dictionaries don't preserve order, we need the 'report_fields' to define the column order.

Open to alternatives.

**Comment by eahrold** [comment]
Right, still waiting on that NSOrderedDictionary to leverage. (Damn hash effecencies!)

This looks good.

Two small things, 
1. It would be nice to have the "failures" array included in the results_report.
1. I did notice that the header is getting defaulted to 'None' https://github.com/autopkg/autopkg/blob/master/Code/autopkg#L1520 which can cause FoundationPlist.py to raise. 

Here's the resulting dict from `autopkg run AdobeFlashPlayer.munki --report-plist=/tmp/rp.plist` with AutoPkg/Cache cleared

```
 {'summary_results': {'url_downloader_summary_result': {'header': None, 'summary_text': 'The following new items were downloaded:', 'data_rows': [u'~/Library/AutoPkg/Cache/local.munki.AdobeFlashPlayer/downloads/AdobeFlashPlayer.dmg']}}}
```

and traceback,

```Python
2015-03-19 21:43:38.799 Python[69172:1107] Property list invalid for format: 100 (property lists cannot contain objects of type 'CFNull')

Traceback (most recent call last):
  File "/mycode/python/autopkg/Code/autopkg", line 1667, in <module>
    sys.exit(main(sys.argv))
  File "/mycode/python/autopkg/Code/autopkg", line 1661, in main
    exit(subcommands[verb]['function'](argv))
  File "/mycode/python/autopkg/Code/autopkg", line 1592, in run_recipes
    write_plist_exit_on_fail(results_report, options.report_plist)
  File "/code/python/autopkg/Code/autopkg", line 535, in write_plist_exit_on_fail
    FoundationPlist.writePlist(plist_dict, path)
  File "/code/python/autopkg/Code/FoundationPlist/FoundationPlist.py", line 128, in writePlist
```

changing it to a simple if:else fixed this for me,
```Python
if type(data).__name__ in ['dict', '__NSCFDictionary']:
    summary_results[key]['header'] = (
        result.get('report_fields') or data.keys())
else:
    summary_results[key]['header'] = []
```

**Comment by gregneagle** [comment]
OrderedDict exists, but was introduced in Python 2.7: https://docs.python.org/2/library/collections.html#collections.OrderedDict

The syntax for creating one is clunky enough that for our purposes we could just use an array of key, value tuples -- instead of 

```python
        self.env["pkg_creator_summary_result"] = {
            'summary_text': 'The following packages were built:',
            'report_fields': ['identifier', 'version', 'pkg_path'],
            'data': {
                'identifier': request['id'],
                'version': request['version'],
                'pkg_path': pkg_path
            }
```

we'd do:

```python
        self.env["pkg_creator_summary_result"] = {
            'summary_text': 'The following packages were built:',
            'data': [
                ('identifier', request['id']),
                ('version', request['version']),
                ('pkg_path', pkg_path)
            ]
```
The array of key, value tuples would keep its order. It might not be harder to work with than a dict...

**Comment by gregneagle** [comment]
My attempt at addressing two issues: https://github.com/autopkg/autopkg/commit/8c972adc760a6f71ec64d318bda1567a677c2ae4

**Comment by gregneagle** [comment]
So: is this fixed and can we close?

**Comment by timsutton** [comment]
I was just looking at this, had a couple ideas:

1. Do you think it might be better for us to be more consistent by having `data_rows` always contain arrays of dicts, rather than dicts for everything except for URLDownloader? I realize we only store a single item for URLDownloader, but for the sake of keeping a consistent structure.
1. I'm wondering if there's also some benefit of autopkg add the recipe identifier as it adds each `data_row` so that whatever is doing the reporting can know what recipe returned what results.

**Comment by timsutton** [comment]
I attempted to address the first idea with 227002293db6661f4b179aabbc70b974dfca2f5f. The change is small, but looks larger because it removes an indented if block.

**Comment by gregneagle** [comment]
>1.  Do you think it might be better for us to be more consistent by having data_rows always contain arrays of dicts, rather than dicts for everything except for URLDownloader? 

And PkgCopier, and Installer, and InstallFromDMG... in fact the simple string is more common than the complex dict.

> I realize we only store a single item for URLDownloader, but for the sake of keeping a consistent structure.

> 2. I'm wondering if there's also some benefit of autopkg add the recipe identifier as it adds each data_row so that whatever is doing the reporting can know what recipe returned what results.

I worry about super-wide report tables, but...
We can actually add as many fields as we like and we don't even have to print all of them! If the headers array exists, only those fields will be printed.

**Comment by timsutton** [comment]
Oh right - I'd forgotten about the Installer processors. I think if the types are rigid it's less conditional logic to carry around (both for us and for anyone using the report plists).

As for 2., it doesn't seem a recipe identifier is so easily accessible from `autopackager`'s results, but maybe I'm missing something. I'll poke a bit further.

**Comment by gregneagle** [comment]
> Oh right - I'd forgotten about the Installer processors. I think if the types are rigid it's less conditional logic to carry around (both for us and for anyone using the report plists).

But now the actual report is a bit clunkier, no? We'd get a header and underline even for simple strings...

**Comment by gregneagle** [comment]
BUT: --report-plist is no longer broken in current master, correct? Can we close this issue?

This isn't really the place for a general implementation discussion...

**Comment by timsutton** [comment]
Sure, let's close.

I agree it looks a little funny to have a heading for a single field, but consider if someone were to write a processor that initially reports a single row item and then they later want to add more fields - the structure of the report would change.

I'm only aware of AutoPkgr and the AutoPkg recipes bot using `--report-plist` so this is all kind of moot, it just seems to me that having a malleable structure is less stable for the parts that move around AutoPkg.

**Comment by gregneagle** [comment]
My initial goal was to replicate the existing summary report formatting as closely as possible, but that might not be terribly important.

**Comment by eahrold** [comment]
Thanks @gregneagle and @timsutton, I can definitely work with this.
